docs/architecture/autofill/generating-fill-scripts.md

```diff
@@ -0,0 +1,217 @@
+
+### Input
+
+The following information is provided to the `generateFillScript()` method:
+
+1. The page details, which represent the information collected from the page in `autofill.js`. It
+   includes:
+   - A list of AutoFillForm objects, representing each Form on the page
+   - A list of AutoFillField objects, representing each Field on the page
+2. The CipherView to fill in on the page
+3. The user's fill options
+
+### Output
+
+The result of the generation routine is a fill script, which is a series of instructions (hence
+"script") to tell the filler what to do to the page.
+
+The script contains an array of instructions with the following types, each of which has a
+corresponding `opid` unique identifier to tie it to an element on the page details:
+
+| Instruction Type | Description                                             |
+| ---------------- | ------------------------------------------------------- |
+| `click_on_opid`  | Click on the element represented by `opid`              |
+| `focus_by_opid`  | Set focus to the element represented by opid            |
+| `fill_by_opid`   | Fill the element represented by opid with value `value` |
+
+The goal of these instructions is to simulate - to the best of our ability - how the page would
+respond to an actual user entering the value into the HTML element. This is why you will see a
+**Click -> Focus -> Fill** sequence often; that simulates what a user would do when entering data.
+
+### Processing
+
+The logic for generating the script has two parts:
+
+- Handle custom fields, and
+- Handle Cipher type-specific fields
+
+For custom fields, the logic is straightforward: the user has defined the field attribute name, so
+we just need to find it in the DOM. If a matching field is found, we find its `opid` and add three
+commands to the script:
+
+1. `click_on_opid`
+2. `focus_by_opid`
+3. `fill_by_opid`
+
+For Cipher type-specific fields, the logic is dependent upon the type. We will look at each one in
+turn below.
+
+#### Login
+
+The core of a `Login` type is a username and password. The easiest field to find is often the
+password, so the routine looks for a password field and then tries to find the most likely candidate
+for a username field that is nearby in the DOM.
+
+For each field in the Page Details collected from the DOM, we check to see if the following
+properties contain the word "password":
+
+- `htmlID`
+- `htmlName`
+- `placeholder`
+
+These will be our "password fields" to fill in on the forms.
+
+Now, we loop through each form on the page. For each of these forms, we examine each password field,
+find the password field on the form, and then try to find the corresponding username field. This is
+done by looking at all fields in the Page Details **prior** to the password in the DOM and finding
+either:
+
+- An exact match from the
+  [UsernameFieldNames](https://github.com/bitwarden/clients/blob/master/apps/browser/src/autofill/services/autofill-constants.ts)
```

Comment:
None of these updated URLs are working for me, sorry! Maybe we just shouldn't try to deep-link to code like this? It'll easily get broken by just renaming a file. We could just refer to `UsernameFieldNames` and let the developer search that in the codebase.

Reply:
The issue here is that one of the autofill-related files was missed with the file migration.  The `autofillConstants` file is still in its old location.  Robyn will be moving it, but for now I've adjusted the link so that it points to the current file location.  When she moves it I'll update it in a separate PR.